Subject: [PATCH] Add a wait for creating crypt block device

Creating the crypt block device takes some time,
especially on low performance platform.
Currently vold isn't aware of it so sometimes vold
tries to access it before the creation is completed.
In that case the encryption or decryption would fail.
To avoid this problem, the wait for creating the block
device is added.

diff --git a/cryptfs.c b/cryptfs.c
index 47acbc3..54aedc4 100644
--- a/cryptfs.c
+++ b/cryptfs.c
@@ -1358,6 +1358,8 @@ static int create_crypto_blk_dev(struct crypt_mnt_ftr *crypt_ftr,
   int version[3];
   char *extra_params;
   int load_count;
+  struct stat statbuf;
+  int retry = 25;
 #ifdef CONFIG_HW_DISK_ENCRYPTION
   char encrypted_state[PROPERTY_VALUE_MAX] = {0};
   char progress[PROPERTY_VALUE_MAX] = {0};
@@ -1436,6 +1438,16 @@ static int create_crypto_blk_dev(struct crypt_mnt_ftr *crypt_ftr,
     SLOGE("Cannot resume the dm-crypt device\n");
     goto errout;
   }
+
+  /* wait for creating dm device */
+  while (stat(crypto_blk_name, &statbuf) != 0) {
+      retry--;
+      if (retry < 0) {
+          SLOGE("Timed out to getting dm blkdev status.\n");
+          goto errout;
+      }
+      usleep(40 * 1000);
+  }
 
   /* We made it here with no errors.  Woot! */
   retval = 0;
